### PR TITLE
codecvt: no conversion for all char-like types of size 1.

### LINKED ISSUE #817: <fstream>: Non-char types are 30 to 50x slower when reading/writing than regular char
**Describe the bug**
Reading and writing to ifstream/ofstream is 30 to 50x slower when using `uint8_t` or `std::byte` vs. using regular `char`.

Attached is a small repro program demonstrating the poor behavior.

This caught me off guard in a project of mine where I attempted to do the "right" thing and was swiftly penalized for it.  I tried to represent a buffer as a `vector<uint8_t>` rather than `vector<char>` and suddenly reading/writing that buffer became enormously slow.  This was just before std::byte became available.  I worked around it using reinterpret_cast<char *>

Now that std::byte is well enough into the ecosystem I decided to test again to see if that's affected as well and it sure is. I can see this becoming a large issue if folks actually embrace the byte in modern, idiomatic c++.

Compiled Version 16.6.0 Preview 6.0

Machine 1:
```
Write to ofstream<char> 0.9732 ms
Write to ofstream<uint8_t> 43.1419 ms
Write to ofstream<std::byte> 45.2306 ms
Read from ifstream<char> 1.0306 ms
Read from ifstream<uint8_t> 43.053 ms
Read from ifstream<std::byte> 43.2853 ms
```

Machine 2:
```
Write to ofstream<char> 0.8127 ms
Write to ofstream<uint8_t> 34.6434 ms
Write to ofstream<std::byte> 34.0428 ms
Read from ifstream<char> 0.9546 ms
Read from ifstream<uint8_t> 32.4374 ms
Read from ifstream<std::byte> 31.9037 ms
```

**Command-line test case**
Build and run the attached vs solution in x64 Release mode

**Expected behavior**
Equivalent performance for all similar data types `char`, `uint8_t`, and `std::byte`

**STL version**
16.6.0 Preview 6.0

[SlowIO.zip](https://github.com/microsoft/STL/files/4605382/SlowIO.zip)

### LINKED ISSUE #2109: <fstream>: Performance issue when reading a binary file using std::basic_ifstream<unsigned char>
**Describe the bug**
Performance issue.
When reading a **binary** file using unsigned data type (ie `std::basic_ifstream<unsigned char>`) is order of magnitude slower than reading it using signed data type (ie `std::basic_ifstream<char>`).
It seems you are doing character by character locale conversion, altough the file is open as binary and not text.

**Command-line test case**
```
Z:\beaus\Desktop\MSVC>type Source.cpp
#include <vector>
#include <iostream>
#include <fstream>
#include <chrono>
#include <filesystem>

int main()
{
        const std::filesystem::path filePath{ "test.data" }; // 128 MB binary file
        const size_t fileSize{ std::filesystem::file_size(filePath) };

        // Read the whole file using char data type
        {
                // Open file in binary mode
                std::basic_ifstream<char> fileStream{ filePath, std::fstream::binary };
                std::vector<char> data;
                data.resize(fileSize);

                const auto start{ std::chrono::system_clock::now() };
                fileStream.read(data.data(), fileSize);
                const auto end{ std::chrono::system_clock::now() };

                std::cout << "Duration read signed: " << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << " ms" << std::endl;
        }

        // Read the whole file using unsigned char data type
        {
                // Open file in binary mode
                std::basic_ifstream<unsigned char> fileStream{ filePath, std::fstream::binary };
                std::vector<unsigned char> data;
                data.resize(fileSize);

                const auto start{ std::chrono::system_clock::now() };
                fileStream.read(data.data(), fileSize);
                const auto end{ std::chrono::system_clock::now() };

                std::cout << "Duration read unsigned: " << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << " ms" << std::endl;
        }

        return 0;
}
Z:\beaus\Desktop\MSVC>cl /EHsc /W4 /WX /std:c++17 .\Source.cpp
Compilateur d'optimisation Microsoft (R) C/C++ version 19.29.30040 pour x64
Copyright (C) Microsoft Corporation. Tous droits réservés.

Source.cpp
Microsoft (R) Incremental Linker Version 14.29.30040.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:Source.exe
Source.obj

Z:\beaus\Desktop\MSVC>Source.exe
Duration read signed: 79 ms
Duration read unsigned: 12838 ms
```

**Expected behavior**
Read a **binary** file using `char` or `unsigned char` should take the same time. (no character conversion for binary data)

**STL version**
* Option 1: Visual Studio version
  * Displayed in Help > About Microsoft Visual Studio
    ```
    Microsoft Visual Studio Community 2019
    Version 16.10.4
    ```

**Additional context**
Use a randomly generated `test.data` binary file to reproduce the issue with my code. Mine is 128MB.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2109
Fixes #817

### CODE REVIEW

**Comment by fsb4000** [comment]
```C++
template <class>
_INLINE_VAR constexpr bool test = false;

template <>
_INLINE_VAR constexpr bool test<char> = true;
```
in two different cpp files gives `fatal error LNK1169: one or more multiply defined symbols found` with clang-cl and C++14.

Fortunately, there is no code like this in the STL.

Ordinary constexpr variables or constexpr templates variables without explicit specialization work fine.

File: stl/inc/xlocale [hunk 1]
```diff
@@ -651,6 +651,49 @@ protected:
...
+#ifdef __cpp_lib_byte
+template <>
+constexpr bool _One_byte_char_like<byte>() {
+    return true;
+}
+#endif // __cpp_lib_byte
+
+#ifdef __cpp_char8_t
+template <>
+constexpr bool _One_byte_char_like<char8_t>() {
```
Comment (AlexGuteniev) [inline stl/inc/xlocale:683]: I'm not sure if specializations is the right thing. Maybe `if constexpr` with `is_same` or `_Is_any_of_v` ?

**Comment by fsb4000** [comment]
Do we need `remove_cv_t<_Ty>` like there: https://github.com/microsoft/STL/blob/60decd0d829e68666b556780002c83605d748d80/stl/inc/xtr1common#L174-L180

**Comment by StephanTLavavej** [comment]
The type traits for the "primary type categories" like `is_integral_v` are required to ignore top-level cv-qualifiers (WG21-N4910  [meta.unary.cat]/2). Our internal type traits don't always `remove_cv_t`, but we need to keep this in mind whenever we're using them, as this is a potential source of bugs.

(`_Is_any_of_v` itself should not `remove_cv_t` as it is a generalized `is_same` which must care about cv-qualifiers; it's things like the "is it safe to activate the `memmeow()` optimization" helpers that sometimes expect usage to have already removed cv.)

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/xlocale [hunk 2]
```diff
@@ -651,6 +651,20 @@ protected:
...
+    byte,
+#endif // __cpp_lib_byte
+#ifdef __cpp_char8_t
+    char8_t,
+#endif // __cpp_char8_t
+    char>;
+
+template <class _Elem, class _Byte>
+_INLINE_VAR constexpr bool _Is_codecvt_do_always_noconv_v =
+    is_same_v<_Byte, _Elem> || (_Is_one_byte_char_like_v<_Byte> && _Is_one_byte_char_like_v<_Elem>);
```
Comment (strega-nil-ms) [inline stl/inc/xlocale:666]: It might be nice to also check for `_Is_two_byte_char_like_v`, for `wchar_t`/`char16_t`?

**Review by strega-nil-ms — APPROVED** [review]
It might be nice to do the same for

File: stl/inc/xlocale (hunk 2 above)
Comment (fsb4000) [reply stl/inc/xlocale:666]: Yes, there is an issue for that: https://github.com/microsoft/STL/issues/605#issuecomment-881972414
but I have not yet explored what needs to be done there to speed up

**Comment by StephanTLavavej** [comment]
Thanks for this major performance improvement! :rocket: :rabbit2: :racing_car:

**Pull request merged**